src/storage/src/object/disk.rs

```diff
@@ -0,0 +1,431 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+
+use bytes::Bytes;
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWriteExt};
+use tokio::sync::mpsc::UnboundedSender;
+
+use crate::object::{BlockLocation, ObjectError, ObjectMetadata, ObjectResult, ObjectStore};
+
+pub(super) mod utils {
+    use std::fs::Metadata;
+    use std::path::Path;
+
+    use tokio::fs::{create_dir_all, File, OpenOptions};
```

Comment:
Just added a simple bench. The bench runs serially in a single thread. The bench result are as followed. Each r/w is operating on 1MB data.
```
fs_operation/tokio      time:   [124.42 ms 126.53 ms 128.76 ms]
tokio stat: Write 1310 times, avg 5.948854961832061ms. Read 1310 times, avg 4.315267175572519ms

fs_operation/std        time:   [27.044 ms 27.805 ms 28.629 ms]
std stat: Write 3270 times, avg 0.0027522935779816515ms. Read 3270 times, avg 0.046788990825688076ms
```

Reply:
> > Do we have plan to support multiple object store backends for Hummock? If we want to spill to disk, I guess we are running two backends. I'm also thinking whether disk should be an object store.
> 
> Yes, in my development plan, we will have two object store, one for local and one for remote. The sstable table store will hold the two object store and do the routing according to the highest bit of the sstable id. A more progressed PR is #2384, and since it's too large, I split in into several smaller PR like this one.
> 
> > Maybe we only need to have disk interface for spill-to-disk, without making it a kind of object store?
> 
> Since the spilled files will be in SST format, making it an object store can share some logic when we read SST and make the code neater. In future PR I will introduce a stream-like uploader for object store so that we don't have to buffer the whole SST in memory when we are using disk object store.

Would we implement it as a hybrid-storage?  Just like following:

```rust
pub struct HybridObjectStore {
    local: Box<dyn ObjectStore>,
    remote: Box<dyn ObjectStore>,
}

impl ObjectStore for HybridObjectStore {
}
```

Reply:
I have thought about it. If we implement it in this way, in the current `ObjectStore` interface, I can't find an elegant way to specify which object store to use. In the future, such `HybridObjectStore` can be used with cache semantic, i.e. we first lookup the local object store, which acts as a local cache, and on cache miss, we then lookup the remote object store.